### PR TITLE
Warn when --iree-llvmcpu-target-cpu defaults to "generic".

### PR DESCRIPTION
**Pull request by bjacob**
Progress on https://github.com/iree-org/iree/issues/18561. This introduces a warning (which we intend to promote to an error in the future) when targeting a generic CPU without explicitly asking for it. This addresses a performance footgun as that IREE default results in low performance.

Along the way this grew into a substantial change to e2e testing rules:
- `TARGET_CPU` and `TARGET_CPU_FEATURES` arguments are gone (were redundant with `COMPILER_FLAGS`).
- For `TARGET_CPU_FEATURES_VARIANTS`, the special value `"default"` is renamed to `"generic"` and a new value `"host"` is also supported.

Example warning (this is customized to the target architecture, here x86):

```
/home/benoit/matmul_i8.mlir:0:0: warning: while creating CPU target: 
Defaulting to targeting a generic CPU for the target architecture will result in poor performance. Please specify a target CPU and/or a target CPU feature set. If it is intended to target a generic CPU, specify "generic" as the CPU.

This can be done in two ways:
1. With command-line flags:
    --iree-llvmcpu-target-cpu=...
    --iree-llvmcpu-target-cpu-features=...
2. Within the IR:
    #hal.executable.target< ... , cpu="...", cpu_features="...">

In the rest of this message, these fields are referred to as just `cpu` and `cpu_features`.

Examples:

    cpu=generic
        Target a generic CPU of the target architecture. The generated code will have poor performance, but will run on any CPU.

    cpu=host
        Target the host CPU. The generated code will have optimal performance on the host CPU but will crash on other CPUs not supporting the same CPU features.

    cpu="name"
        Target a specific CPU. This is mostly used on x86. The accepted values are the same as in Clang command lines.
        List of accepted x86 CPUs: nocona, core2, penryn, bonnell, atom, silvermont, slm, goldmont, goldmont-plus, tremont, nehalem, corei7, westmere, sandybridge, corei7-avx, ivybridge, core-avx-i, haswell, core-avx2, broadwell, skylake, skylake-avx512, skx, cascadelake, cooperlake, cannonlake, icelake-client, rocketlake, icelake-server, tigerlake, sapphirerapids, alderlake, raptorlake, meteorlake, arrowlake, arrowlake-s, lunarlake, gracemont, pantherlake, sierraforest, grandridge, graniterapids, graniterapids-d, emeraldrapids, clearwaterforest, knl, knm, k8, athlon64, athlon-fx, opteron, k8-sse3, athlon64-sse3, opteron-sse3, amdfam10, barcelona, btver1, btver2, bdver1, bdver2, bdver3, bdver4, znver1, znver2, znver3, znver4, znver5, x86-64, x86-64-v2, x86-64-v3, x86-64-v4

    cpu_features="+feature1,..."
        Target a CPU supporting the comma-separated of (+-prefixed) features. The accepted values are the same as in Clang command lines.
```

### CODE REVIEW

**Comment by ScottTodd** [comment]
> WDYT @benvanik @stellaraccident @ScottTodd ?

nit about GitHub: Please put tags like these in comments, not commit messages (or PR descriptions that then become commit messages). If someone pushes a commit containing such a tag to their fork, it will notify those users. I don't really need to know whenever someone rebases or otherwise rewrites history in a fork :P

File: compiler/plugins/target/LLVMCPU/LLVMTargetOptions.cpp [hunk 1]
```diff
@@ -588,6 +512,12 @@ void LLVMCPUTargetCLOptions::bindOptions(OptionsBinder &binder) {
       llvm::cl::cat(category),
       llvm::cl::desc("LLVM target machine CPU features; use 'host' for your "
                      "host native CPU."));
+  binder.opt<std::string>(
+      "iree-llvmcpu-logging-unspecified-target-cpu",
+      loggingUnspecifiedTargetCPU, llvm::cl::cat(category),
+      llvm::cl::desc(
+          "Warning behavior on unspecified target CPU. Accepted values: "
+          " (meaning no logging), \"warn\", \"fatal"));
```
Comment (ScottTodd) [inline compiler/plugins/target/LLVMCPU/LLVMTargetOptions.cpp]: I would use an enum here if possible, rather than a free-form `std::string`. Pretty sure LLVM has a way to do that: https://llvm.org/docs/CommandLine.html.

Or, and I think I like this more, don't add a flag controlling the logging behavior at all. If we roll this warning/error out and see noisy logs and complaining users, _then_ we can tone down the logging.

There are other performance hints that we might want to log, and I don't like the idea of adding a ~40 character long flag to control the logging for each of them. Maybe we could add a top level verbosity setting for diagnostics (or use whatever upstream may provide: https://mlir.llvm.org/docs/Diagnostics/)

See also
* https://github.com/iree-org/iree/issues/13207

File: compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        WARNING: outside of x86, this flag may not actually populate CPU features. It may only serve to select the instruction scheduling model.
+)MSG";
+  }
+
+  if (loggingUnspecifiedTargetCPU == "fatal") {
+    // I know we shouldn't do that in library code. But this is a non-default
+    // flag, and it's going to be very useful to check that we got all instances
+    // fixed before enabling the warning by default.
+    exit(EXIT_FAILURE);
+  }
```
Comment (ScottTodd) [inline compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp]: Yeah, let's not merge with `exit(EXIT_FAILURE)`, even for a non-default flag. We'll need to propagate the error somehow.

File: compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  if (!cpu.empty() || !cpuFeatures.empty()) {
+    return true;
+  }
+  if (loggingUnspecifiedTargetCPU.empty()) {
+    return false;
+  }
+  if (triple.isX86()) {
+    // On X86, one usually talks in terms of CPU, not individual CPU features.
+    llvm::errs() << R"MSG(
+error: Please pass --iree-llvmcpu-target-cpu (or alternatively, --iree-llvmcpu-target-cpu-features, but this message is tailored to the current target architecture, x86, where --iree-llvmcpu-target-cpu is used more often).
```
Comment (ScottTodd) [inline compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp]: Considering the case where the user is compiling a module with these set via attributes and not via CLI flags, this should be an op diagnostic and not go directly to `stderr` via `llvm::errs()`.

Errors may not be as actionable when using a hosting framework that _should_ be setting the target CPU but isn't, so we may still want to just refer to these settings via their CLI flag names.

**Review by ScottTodd — COMMENTED** [review]
Nice! I like the refactoring into the helper file and methods.

We should think through the diagnostics and error handling a bit more, but beyond that I like the new code behavior.

File: compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp (hunk 3 above)
Comment (bjacob) [reply compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp]: All good points! With op diagnostics, can I generate a warning as opposed to an error?

I need to refocus a bit on GPU data tiling now, so we can give this some time to discuss and sort out exactly what we want to do. Maybe we can pick this up when you're back.  Converting back to draft for now.

**Comment by bjacob** [comment]
@ScottTodd , @benvanik , @stellaraccident , I just pushed what I have. It addresses all of @ScottTodd 's earlier review comments, and produces the expected results in `iree-compile` command-line invocations with `target-backends=llvm-cpu`. The main problem that I am struggling with is that it also generates the warning message about the generic GPU fallback... for other target-backends. That is because at the moment, the error message is being printed from `LLVMCPUTargetCLOptions::getTargetOptions`, which is called at target backend registration, and all target backends are registered regardless of which one is actually used.

I don't know how to fix this. @benvanik suggested `buildConfigurationPassPipeline`, but that too is being called for the LLVMCPU backend in a `iree-compile` invocation with `vulkan-spirv`.

**Comment by bjacob** [comment]
After chatting some more with @ScottTodd , pushed a 2nd commit that makes it work AFAICS: https://github.com/iree-org/iree/pull/18682/commits/18748f7bd02e6029e36a56074f9fa6f7a23c4527

WDYT?

**Comment by stellaraccident** [comment]
> After chatting some more with @ScottTodd , pushed a 2nd commit that makes it work AFAICS: [18748f7](https://github.com/iree-org/iree/commit/18748f7bd02e6029e36a56074f9fa6f7a23c4527)
> 
> WDYT?

That seems ok to me.

File: tests/e2e/regression/BUILD.bazel
Comment (ScottTodd) [inline tests/e2e/regression/BUILD.bazel:1]: Can you resolve the merge conflicts on this file (and the CMakeLists.txt), so the github workflows run?

File: build_tools/cmake/iree_check_test.cmake [hunk 4]
```diff
@@ -162,6 +162,9 @@ function(iree_check_test)
   if(_RULE_INPUT_TYPE)
     list(APPEND _BASE_COMPILER_FLAGS "--iree-input-type=${_RULE_INPUT_TYPE}")
   endif()
+  if(_RULE_TARGET_CPU)
+    list(APPEND _BASE_COMPILER_FLAGS "--iree-llvmcpu-target-cpu=${_RULE_TARGET_CPU}")
+  endif()
   if(_RULE_TARGET_CPU_FEATURES)
     list(APPEND _BASE_COMPILER_FLAGS "--iree-llvmcpu-target-cpu-features=${_RULE_TARGET_CPU_FEATURES}")
   endif()
```
Comment (ScottTodd) [inline build_tools/cmake/iree_check_test.cmake]: What's the rationale for these getting their own flags instead of using `COMPILER_FLAGS`?

In general, I want these build system functions to do _less_, not _more_ - especially for global flags controlling target backend/device-specific options.

I don't really mind being this explicit in our tests:
```cmake
  TARGET_BACKEND
    "llvm-cpu"
  COMPILER_FLAGS
    "--iree-llvmcpu-target-cpu=generic"
  DRIVER
    "local-task"
```

File: build_tools/cmake/iree_e2e_generated_runner_test.cmake [hunk 5]
```diff
@@ -48,10 +48,15 @@ function(iree_e2e_runner_test)
...
     "NAME;TEST_TYPE;VARIANT_NAME;TESTS_SRC;TESTS_VMFB;CALLS_SRC;CALLS_VMFB;TRACE;TARGET_BACKEND;DRIVER;TEST_RUNNER;TEST_DEFINED;TEST_DISABLED"
-    "COMPILER_FLAGS;RUNNER_ARGS;LABELS;TARGET_CPU_FEATURES"
+    "COMPILER_FLAGS;RUNNER_ARGS;LABELS;TARGET_CPU;TARGET_CPU_FEATURES"
     ${ARGN}
   )
 
+  if (_RULE_NAME MATCHES "e2e_attention_cpu_f16_f16_f16_small")
+    message(STATUS "_RULE_TARGET_CPU=${_RULE_TARGET_CPU}")
+    message(STATUS "_RULE_TARGET_CPU_FEATURES=${_RULE_TARGET_CPU_FEATURES}")
+  endif()
```
Comment (ScottTodd) [inline build_tools/cmake/iree_e2e_generated_runner_test.cmake]: Remove debug logging before merge

File: compiler/plugins/target/LLVMCPU/LLVMTargetOptions.cpp [hunk 6]
```diff
@@ -333,9 +259,18 @@ LLVMTarget::loadFromConfigAttr(Location loc, DictionaryAttr config,
...
       return {};
     }
+    ResolveCPUAndCPUFeaturesStatus status;
     std::optional<LLVMTarget> maybeTarget =
-        LLVMTarget::create(*triple, cpu ? *cpu : "generic",
-                           cpuFeatures ? *cpuFeatures : "", linkEmbedded);
+        LLVMTarget::create(*triple, cpu.value_or(""), cpuFeatures.value_or(""),
+                           linkEmbedded, status);
+    if (status != ResolveCPUAndCPUFeaturesStatus::OK) {
+      // TODO(bjacob): after people have had time to adapt, typically by adding
```
Comment (ScottTodd) [inline compiler/plugins/target/LLVMCPU/LLVMTargetOptions.cpp]: Link the issue (https://github.com/iree-org/iree/issues/18561) in the TODO

```suggestion
      // TODO(#18561): after people have had time to adapt, typically by adding
```

File: compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp [hunk 7]
```diff
@@ -0,0 +1,176 @@
+// Copyright 2020 The IREE Authors
```
Comment (ScottTodd) [inline compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp]: Mostly new file can use 2024 instead of 2020

File: compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+  switch (status) {
+  case ResolveCPUAndCPUFeaturesStatus::ImplicitGenericFallback:
+    return getImplicitGenericFallbackMessage(triple);
+  case ResolveCPUAndCPUFeaturesStatus::InconsistentHost:
+    return "If either CPU or CPU-features is `host`, the other must "
+           "be either also `host` or the default value.\n";
+  case ResolveCPUAndCPUFeaturesStatus::UnimplementedMapping:
+    return "Resolution of CPU to CPU-features is not implemented on this "
+           "target architecture. Pass explicit "
+           "CPU-features, or implement the missing mapping.";
```
Comment (ScottTodd) [inline compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp]: Include `\n` consistently?

```suggestion
    return "If either CPU or CPU-features is `host`, the other must "
           "be either also `host` or the default value.\n";
  case ResolveCPUAndCPUFeaturesStatus::UnimplementedMapping:
    return "Resolution of CPU to CPU-features is not implemented on this "
           "target architecture. Pass explicit "
           "CPU-features, or implement the missing mapping.\n";
```

File: compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+                      std::string &cpuFeatures) {
+  if (triple.isAArch64()) {
+    llvm::SubtargetFeatures targetCpuFeatures(cpuFeatures);
+    // x18 is platform-reserved per the Aarch64 procedure call specification.
+    targetCpuFeatures.AddFeature("reserve-x18", true);
+    cpuFeatures = targetCpuFeatures.getString();
+  }
+}
+
+std::string getImplicitGenericFallbackMessage(const llvm::Triple &triple) {
```
Comment (ScottTodd) [inline compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp]: This message is great, thank you!

**Review by ScottTodd — COMMENTED** [review]
Nice! No major concerns on my end with this approach.

File: build_tools/cmake/iree_check_test.cmake (hunk 4 above)
Comment (bjacob) [reply build_tools/cmake/iree_check_test.cmake]: Good catch! I just made that change in all related rules.

**Comment by bjacob** [comment]
@ScottTodd , this feels ready for review now.

I have followed your suggestion to drop `TARGET_CPU` and `TARGET_CPU_FEATURES`. While doing do, I found the place which had originally motivated them: this is where we pass a `--requirements` flag to the test runner so that it can check at runtime that the host CPU supports the expected features. Now that there isn't a separate `TARGET_CPU_FEATURES` field anymore, we resort to parsing the `COMPILER_FLAGS`, which is not too bad and is a tiny detail, so it is the right call to deal with it rather than have those `TARGET_CPU` and `TARGET_CPU_FEATURES` flags everywhere.

The `TARGET_CPU_FEATURES_VARIANTS` field used to accept a special value `"default"` and default to that. I renamed that to `"generic"` because that's what it does (so let's not have two different names for that same configuration, `"default"` and `"generic"`) and I added support for `"host"` so we can easily add tests that test host codegen and/or decide to default to testing that in addition to `"generic"` in a follow-up.

As part of this CL, I had to add `"--iree-llvmcpu-target-cpu=generic"` to a number of `iree_check_single_backend_test_suite`.  If those were `iree_check_test_suite`, this wouldn't be needed, as those take `TARGET_CPU_FEATURES_VARIANTS` which default to `"generic"`. This discrepancy between `iree_check_single_backend_test_suite` and `iree_check_test_suite` seems unfortunate and not reflected in the names. We should address that in a follow-up. My preferred resolution actually would be to drop `iree_check_single_backend_test_suite` other than as a local implementation helper within `iree_check_test.cmake`. Everyone should be able to just use `iree_check_test_suite`, and it's not even more verbose.

**Comment by ScottTodd** [comment]
> I have followed your suggestion to drop `TARGET_CPU` and `TARGET_CPU_FEATURES`. While doing do, I found the place which had originally motivated them: this is where we pass a `--requirements` flag to the test runner so that it can check at runtime that the host CPU supports the expected features. Now that there isn't a separate `TARGET_CPU_FEATURES` field anymore, we resort to parsing the `COMPILER_FLAGS`, which is not too bad and is a tiny detail, so it is the right call to deal with it rather than have those `TARGET_CPU` and `TARGET_CPU_FEATURES` flags everywhere.

Ah. That's on my list to remove too: https://github.com/iree-org/iree-test-suites/blob/3a0ea13cbdc954365d653a48cc99cc63a9ff09b3/linalg_ops/matmul/generate_e2e_matmul_tests.py#L529-L531
```
    # TODO(scotttodd): drop this and whatever logic in the test tool used it
    #     multiple backends should be able to use the same input IR, so the
    #     input IR shouldn't need things like CPU features in it
```

Also for code review, please try to avoid force pushing. I don't see a way to view the diff between when I last reviewed and the latest code >_>

**Comment by ScottTodd** [comment]
(GitHub has a few ways to view diffs between commits, but force pushing with a rebase defeats most of them)
* Changes since last review shows no diff 
  image
* "Compare" on the force push shows _all_ changes, including commits unrelated to the PR
  image

**Comment by bjacob** [comment]
@ScottTodd , for my education - when my PR has a conflict with `main`, so I have to rebase it, that in itself seems to require a force-push, right  -- so do I correctly understand your suggestion in such cases to refrain from squashing the commits in this PR, so that the force-push of the rebased first commit doesn't prevent you from looking at the other commits on top of it?

**Comment by ScottTodd** [comment]
> @ScottTodd , for my education - when my PR has a conflict with `main`, so I have to rebase it, that in itself seems to require a force-push, right -- so do I correctly understand your suggestion in such cases to refrain from squashing the commits in this PR, so that the force-push of the rebased first commit doesn't prevent you from looking at the other commits on top of it?

`git merge upstream/main` then resolve conflicts should avoid the need to rebase or force push.

**Comment by bjacob** [comment]
TIL, thanks.  Because we ultimately rebase-and-squash when merging PRs, I had followed that workflow also locally. It hadn't occurred to me that I could create merge commits locally and on my PR branch, as that would still ultimately get rebased-and-squashed on merging.

File: build_tools/cmake/iree_hal_cts_test_suite.cmake [hunk 10]
```diff
@@ -87,6 +87,9 @@ function(iree_hal_cts_test_suite)
       "--iree-hal-target-backends=${_RULE_COMPILER_TARGET_BACKEND}"
       ${_RULE_COMPILER_FLAGS}
     )
+    if(_RULE_COMPILER_TARGET_BACKEND STREQUAL "llvm-cpu")
+      list(APPEND _TRANSLATE_FLAGS "--iree-llvmcpu-target-cpu=generic")
+    endif()
```
Comment (ScottTodd) [inline build_tools/cmake/iree_hal_cts_test_suite.cmake]: Add to `COMPILER_FLAGS` at the call sites instead of adding a branch to the common function

* https://github.com/iree-org/iree/blob/main/runtime/src/iree/hal/drivers/local_sync/cts/CMakeLists.txt
* https://github.com/iree-org/iree/blob/main/runtime/src/iree/hal/drivers/local_task/cts/CMakeLists.txt

File: build_tools/cmake/iree_e2e_generated_runner_test.cmake [hunk 11]
```diff
@@ -251,9 +243,13 @@ function(iree_single_backend_e2e_runner_test)
...
-  if(_RULE_TARGET_CPU_FEATURES)
-    list(APPEND _GENERATOR_STANDARD_FLAGS "--requirements=${_RULE_TARGET_CPU_FEATURES}")
-  endif()
+  foreach(_COMPILER_FLAG IN LISTS _RULE_COMPILER_FLAGS)
+    set(_CPU_FEATURES_REGEX "^--iree-llvmcpu-target-cpu-features=")
+    if (_COMPILER_FLAG MATCHES "${_CPU_FEATURES_REGEX}")
+      string(REGEX REPLACE "${_CPU_FEATURES_REGEX}" "" _CPU_FEATURES "${_COMPILER_FLAG}")
+      list(APPEND _GENERATOR_STANDARD_FLAGS "--requirements=${_CPU_FEATURES}")
+    endif()
+  endforeach()
```
Comment (ScottTodd) [inline build_tools/cmake/iree_e2e_generated_runner_test.cmake:252]: This is fine for now / in this repo.

> > > I have followed your suggestion to drop `TARGET_CPU` and `TARGET_CPU_FEATURES`. While doing do, I found the place which had originally motivated them: this is where we pass a `--requirements` flag to the test runner so that it can check at runtime that the host CPU supports the expected features. Now that there isn't a separate `TARGET_CPU_FEATURES` field anymore, we resort to parsing the `COMPILER_FLAGS`, which is not too bad and is a tiny detail, so it is the right call to deal with it rather than have those `TARGET_CPU` and `TARGET_CPU_FEATURES` flags everywhere.
> 
> Ah. That's on my list to remove too: https://github.com/iree-org/iree-test-suites/blob/3a0ea13cbdc954365d653a48cc99cc63a9ff09b3/linalg_ops/matmul/generate_e2e_matmul_tests.py#L529-L531
> 
> ```
>     # TODO(scotttodd): drop this and whatever logic in the test tool used it
>     #     multiple backends should be able to use the same input IR, so the
>     #     input IR shouldn't need things like CPU features in it
> ```

(Over in [iree-test-suites](https://github.com/iree-org/iree-test-suites)) I think I also want to push this logic to the leaves, similar to how we have `if(IREE_HIP_TEST_TARGET_CHIP)` controlling defining and running tests for the ROCm backend + HIP driver, we can check explicit options like `IREE_TEST_SUITES_INCLUDE_ARM_SME_TESTS` or check automatically/programmatically for support.

Basically, each set of tests should make it obvious how the tests are enabled. The existing logic, which you are updating here, hides some of that decision making in helper functions. For remote targets like Android devices and local targets like GPUs, we can't as easily enumerate devices and supported features at configure time, and I don't want to special case any backend in the helper functions.

File: build_tools/bazel/iree_check_test.bzl [hunk 12]
```diff
@@ -43,19 +42,16 @@ def iree_check_test(
...
 
-    if target_cpu_features:
-        fail("target_cpu_features must currently be empty")
     input_type_flags = []
     if input_type:
         input_type_flags = ["--iree-input-type=%s" % input_type]
     flags = [
         "--iree-hal-target-backends=%s" % target_backend,
+        "--iree-llvmcpu-target-cpu=generic",
     ] + compiler_flags + input_type_flags
```
Comment (ScottTodd) [inline build_tools/bazel/iree_check_test.bzl]: Is this implicit flag needed in the Bazel helper function? It looks like you already updated several call sites to pass it explicitly. We shouldn't always set a `llvmcpu`, even for tests using other targets.

I don't care as much about the Bazel side of that though, and it looks like CMake has the behavior that I want already.

File: build_tools/bazel/iree_e2e_generated_runner_test.bzl [hunk 13]
```diff
@@ -44,23 +43,18 @@ def iree_e2e_runner_test(
...
         name = name + "_%s_module" % test_type,
         module_name = tests_vmfb,
         src = tests_src,
         flags = [
             "--iree-hal-target-backends=%s" % target_backend,
-        ] + ([
-            "--iree-llvmcpu-target-cpu-features=%s" % target_cpu_features,
-        ] if target_cpu_features else []) + compiler_flags,
+            "--iree-llvmcpu-target-cpu=generic",
+        ] + compiler_flags,
```
Comment (ScottTodd) [inline build_tools/bazel/iree_e2e_generated_runner_test.bzl]: Same comment here about implicit flags in the helper function vs explicit flags at the call sites

**Review by ScottTodd — COMMENTED** [review]
Mostly LGTM. Just a few lingering comments.

File: build_tools/cmake/iree_hal_cts_test_suite.cmake (hunk 10 above)
Comment (bjacob) [reply build_tools/cmake/iree_hal_cts_test_suite.cmake]: done.

File: build_tools/bazel/iree_check_test.bzl (hunk 12 above)
Comment (bjacob) [reply build_tools/bazel/iree_check_test.bzl]: Good catch. There was something to change here, but to match what we do in the CMake implementation, it had to be something else. The top-level rules like `iree_check_test_suite` and `iree_generated_e2e_runner_test` take the special `target_cpu_features_variants` parameter. This defaults to `["generic"]` on CMake, so it should also default to that on Bazel. So the `"generic"` stays implicit but moves one level up within the rules.

File: build_tools/bazel/iree_e2e_generated_runner_test.bzl (hunk 13 above)
Comment (bjacob) [reply build_tools/bazel/iree_e2e_generated_runner_test.bzl]: same.

**Review by ScottTodd — APPROVED** [review]

**Comment by bjacob** [comment]
@stellaraccident , i'm blissfully unaware of release schedules. If there is any upcoming release, this one is worth having on it.

**Comment by ScottTodd** [comment]
Build failures on MSVC: https://github.com/iree-org/iree/actions/runs/11401187470/job/31723569010#step:7:7793

> @stellaraccident , i'm blissfully unaware of release schedules. If there is any upcoming release, this one is worth having on it.

See the pinned issue: https://github.com/iree-org/iree/issues/18432

**Comment by bjacob** [comment]
@ScottTodd 

> Build failures on MSVC: https://github.com/iree-org/iree/actions/runs/11401187470/job/31723569010#step:7:7793

https://github.com/iree-org/iree/pull/18830

**Comment by ScottTodd** [comment]
> > @stellaraccident , i'm blissfully unaware of release schedules. If there is any upcoming release, this one is worth having on it.
> 
> See the pinned issue: #18432

@bjacob I added a bullet to the rolling release notes there:

> * Added warnings for when `--iree-llvmcpu-target-cpu=` or `--iree-llvmcpu-target-cpu-features` are missing, since the default of "generic" is slow. These flags will be required in the future: [0c6a151](https://github.com/iree-org/iree/commit/0c6a151c65285987f5daabc7f76fe57a82b45ab1)

When we next cut a stable release, the release notes on that issue will be included. Feel free to edit the issue if you have other phrasing you want us to use.
